### PR TITLE
lsm: remove compaction interface type

### PR DESCRIPTION
**Pull request by matklad**
Use (tree_id, level_b) index pair for "dynamic" dispatch instead

### CODE REVIEW

**Comment by cb22** [comment]
@matklad not sure if you saw the failure here - the fuzz needs to be changed too.

File: src/lsm/forest.zig [hunk 1]
```diff
@@ -921,6 +956,9 @@ fn CompactionPipelineType(comptime Forest: type, comptime Grid: type) type {
             const first_beat = compaction_beat == 0;
             const half_beat = compaction_beat == @divExact(constants.lsm_batch_multiple, 2);
 
+            if (self.forest == null) self.forest = forest;
```
Comment (matklad) [inline src/lsm/forest.zig:959]: Forest does `bar_apply_to_manifest` from `compact_callback`, so I figured it'd be simpler to just always have forest set. We still have `callback` to assert the proper lifecycle. 

Though, I do not like re-entrancy of `Forest.compact_callback`, would be good to fix that, but out of scope for this PR.

File: src/lsm/forest.zig [hunk 2]
```diff
@@ -1060,35 +1102,18 @@ fn CompactionPipelineType(comptime Forest: type, comptime Grid: type) type {
...
-            self.forest = null;
 
             callback(forest);
         }
 
         // TODO: It would be great to get rid of *anyopaque here. Batiati's scan approach
         // wouldn't compile for some reason.
         fn blip_callback(
-            compaction_interface_opaque: *anyopaque,
+            slot_opaque: *anyopaque,
```
Comment (matklad) [inline src/lsm/forest.zig:1112]: Looks like just passing a pointer to a slot works here?

File: src/testing/cluster/manifest_checker.zig [hunk 3]
```diff
@@ -54,15 +54,16 @@ pub fn ManifestCheckerType(comptime Forest: type) type {
             for (0..constants.lsm_levels) |level| {
                 checksum_stream.add(std.mem.asBytes(&level));
 
-                inline for (Forest.tree_id_range.min..Forest.tree_id_range.max + 1) |tree_id| {
+                inline for (Forest.tree_id_range.min..Forest.tree_id_range.max + 1) |tree_id_u16| {
```
Comment (matklad) [inline src/testing/cluster/manifest_checker.zig:57]: Intentionally avoiding `enum.values` here such that the verification code doesn't rely on the order of filds in an enum, and does a stupid numeric loop.

**Comment by matklad** [comment]
ready for review @cb22 !

File: src/lsm/forest.zig (hunk 1 above)
Comment (cb22) [reply src/lsm/forest.zig:959]: > Though, I do not like re-entrancy of Forest.compact_callback, would be good to fix that, but out of scope for this PR.

Yes - that was originally to get things working with the existing manifest compaction without having to touch that code too!

File: src/lsm/forest.zig [hunk 4]
```diff
@@ -1268,11 +1332,26 @@ fn CompactionPipelineType(comptime Forest: type, comptime Grid: type) type {
 
                 // TODO: CompactionInterface internally stores a pointer to the real
```
Comment (cb22) [inline src/lsm/forest.zig]: Dead comment here

**Review by cb22 — COMMENTED** [review]
LGTM - just a dangling old comment!

File: src/lsm/forest.zig (hunk 4 above)
Comment (matklad) [reply src/lsm/forest.zig]: Thanks, removed!

**Review by cb22 — APPROVED** [review]

**Pull request merged**